kotlin/show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgisruntime/sample/showdevicelocationusingindoorpositioning/MainActivity.kt

```diff
@@ -0,0 +1,418 @@
+        textView.text = locationPropertiesMessage
+    }
+
+    /**
+     * Set up the floor layer when the device location is updated
+     */
+    private fun setupLayers() {
+        // update layer's definition express with the current floor
+        val layerList: LayerList = mapView.map.operationalLayers
+        for (layer in layerList) {
+            val name = layer.name
+            if (layer is FeatureLayer && (name == "Details" || name == "Units" || name == "Levels")) {
+                layer.definitionExpression = "VERTICAL_ORDER = $mCurrentFloor"
+            }
+        }
+    }
+
+
+    /**
+     * Handle status changes of IndoorsLocationDataSource using the [statusChangedEvent]
+     */
+    override fun statusChanged(statusChangedEvent: LocationDataSource.StatusChangedEvent?) {
+        when (statusChangedEvent?.status) {
+            LocationDataSource.Status.STARTING -> progressBar.visibility = View.VISIBLE
+            LocationDataSource.Status.STARTED -> progressBar.visibility = View.GONE
+            LocationDataSource.Status.FAILED_TO_START -> {
+                progressBar.visibility = View.GONE
+                Log.e(TAG, "Failed to start IndoorsLocationDataSource")
+                Toast.makeText(this, "Failed to start IndoorsLocationDataSource", Toast.LENGTH_LONG).show()
+            }
+            LocationDataSource.Status.STOPPED -> {
+                progressBar.visibility = View.GONE
+                stopLocationDisplay()
+                Log.e(TAG, "IndoorsLocationDataSource stopped due to an internal error")
+                Toast.makeText(this, "IndoorsLocationDataSource stopped due to an internal error", Toast.LENGTH_LONG).show()
+            }
+        }
+    }
+
+    /**
+     * Check for location permissions, if not received then request for one
+     */
+    private fun checkPermissions() {
+        val requestCode = 1
+        val requestPermissions = arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
+        if (ContextCompat.checkSelfPermission(this, requestPermissions[0]) != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this, requestPermissions, requestCode)
+        } else {
+            // permission already given, so no need to request
+            connectToPortal()
+        }
+    }
+
+    /**
+     * Result of the user from location permissions request
+     */
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String?>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            // if location permissions accepted, start setting up IndoorsLocationDataSource
+            connectToPortal()
+        } else {
+            val message = "Location permission is not granted"
+            Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
+            Log.e(TAG, message)
```

Comment:
Do we need to ask the user again for permission if they don't grant it?

Reply:
Permissions are requested on app launch. If the user would deny it more than twice (Android behavior), the permission dialog wouldn't show. Instead a toast message "Location permission is not granted" would be shown. Asking again wouldn't benefit as the Android would suppress the permission